### PR TITLE
Update Switch

### PR DESCRIPTION
**Pull request by jaortiz**

### CODE REVIEW

File: components/switch/examples/90-local-tokens.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import { useState } from 'react';
+import { GEL, jsx } from '@westpac/core';
+import { Switch } from '@westpac/switch';
+
+function Example({ brand }) {
+	const [checked, setChecked] = useState(false);
+	const brandWithTokens = { ...brand };
+
+	brandWithTokens['@westpac/switch'] = {
+		labelCss: {
```
Comment (dominikwilkowski) [inline components/switch/examples/90-local-tokens.js]:
```suggestion
		labelCSS: {
```

File: components/switch/examples/90-local-tokens.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+			padding: '10px',
+		},
+		textCss: {
```
Comment (dominikwilkowski) [inline components/switch/examples/90-local-tokens.js]:
```suggestion
		textCSS: {
```

File: components/switch/examples/90-local-tokens.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+			color: 'palevioletred',
+		},
+		toggleCss: {
```
Comment (dominikwilkowski) [inline components/switch/examples/90-local-tokens.js]:
```suggestion
		toggleCSS: {
```

File: components/switch/examples/90-local-tokens.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+			borderColor: 'mediumVioletred',
+			backgroundColor: checked ? 'paleVioletRed' : 'white',
+		},
+		toggleTextCss: { color: checked ? 'white' : 'firebrick' },
```
Comment (dominikwilkowski) [inline components/switch/examples/90-local-tokens.js]:
```suggestion
		toggleTextCSS: { color: checked ? 'white' : 'firebrick' },
```

File: components/switch/src/Switch.js [hunk 5]
```diff
@@ -1,28 +1,14 @@
...
-	const context = useContext(SwitchContext);
-	if (!context) {
-		throw new Error('Switch sub-components should be wrapped in a <Switch>.');
-	}
-	return context;
-};
-
+import { jsx, useBrand, useMediaQuery, merge } from '@westpac/core';
+import { VisuallyHidden } from '@westpac/a11y';
+import shortid from 'shortid';
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: I don't think `shortid` is save to use. We should add the sequential id to `document` and just iterate it.

File: components/switch/src/Switch.js [hunk 6]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+}) => {
+	const { COLORS, [pkg.name]: localBrandTokens } = useBrand();
+	const mq = useMediaQuery();
+	const [checked, setChecked] = useState(isChecked);
+	const [switchId] = useState(`switch-${shortid.generate()}`);
+
+	const sizeArr = responsiveMap(asArray(size));
+
+	const localTokens = {
+		labelCss: {},
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
		labelCSS: {},
```

File: components/switch/src/Switch.js [hunk 7]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+		textCss: {},
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
		textCSS: {},
```

File: components/switch/src/Switch.js [hunk 8]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+		toggleCss: {},
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
		toggleCSS: {},
```

File: components/switch/src/Switch.js [hunk 9]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+		toggleTextCss: {},
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
		toggleTextCSS: {},
```

File: components/switch/src/Switch.js [hunk 10]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+	name,
+	label,
+	checked: isChecked,
+	onChange,
+	size,
+	block,
+	flipped,
+	toggleText,
+	disabled,
+	srOnlyText,
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: We should still spread the rest:

```suggestion
	srOnlyText,
	...rest,
```

File: components/switch/src/Switch.js [hunk 11]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
 
-export const Switch = ({ size, ...props }) => {
-	const flexiSize = responsiveMap(asArray(size));
+	useEffect(() => {
+		setChecked(isChecked);
+	}, [isChecked]);
+
+	const handleChange = name => () => {
+		if (onChange) {
+			onChange(name);
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: We should use `wrapHandlers` from core and combine the consumer `onChange` with out internal state change.

File: components/switch/src/Switch.js [hunk 12]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+			setChecked(!checked);
+		}
+	};
 
 	return (
-		<SwitchContext.Provider value={{ flexiSize }}>
-			<SwitchLabel {...props} />
-		</SwitchContext.Provider>
+		<label
+			htmlFor={switchId}
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: You don't need to use and id or `for` attribute if you wrap the input inside a label (which you're doing here) and avoid using unique IDs all-together.

Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
```

File: components/switch/src/Switch.js [hunk 13]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+				marginRight: !block && '1.125rem',
+				height: !block && sizeArr.height,
+				marginBottom: '0.375rem',
+				flexDirection: flipped && 'row-reverse',
+				cursor: disabled ? 'not-allowed' : 'pointer',
+				...localTokens.labelCss,
+			})}
+		>
+			<input
+				id={switchId}
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
```

File: components/switch/src/Switch.js [hunk 14]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+				width: block && '100%',
+				flexWrap: 'wrap',
+				alignItems: 'center',
+				position: 'relative',
+				marginRight: !block && '1.125rem',
+				height: !block && sizeArr.height,
+				marginBottom: '0.375rem',
+				flexDirection: flipped && 'row-reverse',
+				cursor: disabled ? 'not-allowed' : 'pointer',
+				...localTokens.labelCss,
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: This is where we should spread `rest` and we don't need `labelCSS` really.
```suggestion
```

File: components/switch/src/Switch.js [hunk 15]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+			})}
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
			})}
		{...rest}
```

File: components/switch/src/Switch.js [hunk 16]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+			/>
+			<span
+				css={{
+					flex: block && 1,
+					display: 'flex',
+					alignItems: 'center',
+					whiteSpace: 'normal',
+					position: 'relative',
+					[flipped ? 'paddingLeft' : 'paddingRight']: '0.375rem',
+					...localTokens.textCss,
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
					...localTokens.textCSS,
```

File: components/switch/src/Switch.js [hunk 17]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+				checked={checked}
+				onChange={handleChange(name)}
+				disabled={disabled}
+				css={{
+					position: 'absolute',
+					zIndex: '-1',
+					opacity: 0,
+				}}
+			/>
+			<span
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: This should be a token in-case someone wan'ts so use their own component.

File: components/switch/src/Switch.js [hunk 18]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+						position: 'absolute',
+						left: checked ? '100%' : 0,
+						transform: checked ? 'translateX(-100%)' : null,
+						top: 0,
+						borderRadius: '50%',
+						backgroundColor: '#fff',
+						boxShadow: '3px 0 6px 0 rgba(0,0,0,0.3)',
+						transition: 'all .3s ease',
+					},
+					...localTokens.toggleCss,
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
					...localTokens.toggleCSS,
```

File: components/switch/src/Switch.js [hunk 19]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+				})}
+			>
+				{!!toggleText.length && (
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: Wait... a boolean length? Why not just `!!toggleText`?

File: components/switch/src/Switch.js [hunk 20]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+					<>
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: We should use `<Fragment>` to make it explicit and avoid issues with syntax highlighter

Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
					<Fragment>
```

File: components/switch/src/Switch.js [hunk 21]
```diff
@@ -62,14 +48,143 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+						</ToggleText>
+						<ToggleText
+							position={'right'}
+							checked={!checked}
+							size={sizeArr}
+							css={localTokens.toggleTextCss}
+						>
+							{toggleText[1]}
+						</ToggleText>
+					</>
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
					</Fragment>
```

File: components/switch/src/Switch.js [hunk 22]
```diff
@@ -1,28 +1,14 @@
 /** @jsx jsx */
 
-import React, { createContext, useContext } from 'react';
+import React, { useState, useEffect } from 'react';
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]:
```suggestion
import React, { useState, useEffect, Fragment } from 'react';
```

File: components/switch/src/Switch.js [hunk 23]
```diff
@@ -62,14 +48,125 @@ const asArray = val => (Array.isArray(val) ? val : [val]);
...
+		Label,
+	};
+
+	merge(overwrites, overwritesWithTokens);
+
+	useEffect(() => {
+		setChecked(isChecked);
+	}, [isChecked]);
+
+	const handleChange = () => wrapHandlers(onChange, () => setChecked(!checked));
```
Comment (jaortiz) [inline components/switch/src/Switch.js]: @dominikwilkowski I'm unsure if this is the correct use of wrapHandlers. From looking at the wrap handlers code, we run both the passed in consumerHandler and our own internal handler unless specified in the consumerHandler via preventDefault() ?

Comment (dominikwilkowski) [reply components/switch/src/Switch.js]: That looks absolutely correct :)
For more infos look at https://github.com/WestpacGEL/GEL/blob/develop/components/core/examples/90-wrapHandlers.js?ts=2

File: components/switch/src/Switch.js [hunk 24]
```diff
@@ -160,28 +143,28 @@ export const Switch = ({
...
 						transition: 'all .3s ease',
 					},
-					...localTokens.toggleCss,
+					...overwrites.toggleCSS,
 				})}
 			>
-				{!!toggleText.length && (
-					<>
+				{!!toggleText && (
+					<Fragment>
```
Comment (dominikwilkowski) [inline components/switch/src/Switch.js]: should probably also be in the overwrites?

File: components/switch/examples/90-local-tokens.js [hunk 25]
```diff
@@ -4,33 +4,40 @@ import { useState } from 'react';
 import { GEL, jsx } from '@westpac/core';
 import { Switch } from '@westpac/switch';
 
+const LabelNew = props => (
```
Comment (dominikwilkowski) [inline components/switch/examples/90-local-tokens.js]: Why not just `Label`?

**Pull request merged**